### PR TITLE
refactor(sandbox-host): remove double wasmer module serialization

### LINKED ISSUE #4050: Remove double `wasmer::Module` serialization when writing to caches
### File Location(s)

See `sandbox/host/src/sandbox/wasmer_backend.rs`

### Proposal

Follow up after Wasmer version upgrade (v4).
After Wasmer version upgrade there is a problem with double `wasmer::Module` serialize when writing to caches.

Relates to #4037

Implement custom LRU + FS cache to eliminate double serialization problem.

### PR DESCRIPTION
**Pull request by ByteNacked**
Resolves #4050 .

Implements custom file system cache to eliminate double serialization problem.

@reviewer-or-team

### CODE REVIEW

File: sandbox/host/src/sandbox/wasmer_backend/cache.rs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        let _ = fs::remove_file(path);
+    }
+
+    /// Store the serialized module in the cache.
+    fn store(&mut self, key: Hash, serialized_module: &[u8]) -> Result<(), io::Error> {
+        let filename = if let Some(ref ext) = self.ext {
+            format!("{}.{}", key, ext)
+        } else {
+            key.to_string()
+        };
```
Comment (gshep) [inline sandbox/host/src/sandbox/wasmer_backend/cache.rs:203]: extract standalone utility function

```suggestion
fn compose_filename(extension: Option<&String>, key: String) -> String {
     match extension {
         Some(extension) => format!("{key}.{extension}"),
         None => key,
      }
}

        let filename = compose_filename(self.ext.as_ref(), key.to_string());
```

File: sandbox/host/src/sandbox/wasmer_backend/cache.rs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    /// so delete the cache file
+    fn remove_key(&self, key: Hash) {
+        let filename = if let Some(ref ext) = self.ext {
+            format!("{}.{}", key, ext)
+        } else {
+            key.to_string()
+        };
+        let path = self.path.join(filename);
+
+        let _ = fs::remove_file(path);
```
Comment (gshep) [inline sandbox/host/src/sandbox/wasmer_backend/cache.rs:194]: would be nice to log the result

File: sandbox/host/src/sandbox/wasmer_backend/cache.rs [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+    // Store module in LRU cache
+    let res = lru_cache()
+        .lock()
+        .expect("CACHED_MODULES lock fail")
+        .insert(CachedModule {
+            wasm: wasm.to_vec(),
+            serialized_module: serialized_module.clone(),
+        });
+    log::trace!("Store module in LRU cache with result: {:?}", res.is_some());
```
Comment (gshep) [inline sandbox/host/src/sandbox/wasmer_backend/cache.rs:119]: maybe just like that?
```suggestion
    log::trace!("Store module in LRU cache with result: {res:?}");
```

File: sandbox/host/src/sandbox/wasmer_backend/cache.rs [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        .path();
+    log::trace!("Wasmer sandbox cache dir is: {cache_path:?}");
+
+    FileSystemCache::new(cache_path).map_err(|_| FileSystemErr)
+}
+
+pub fn get_cached_module(
+    wasm: &[u8],
+    store: &sandbox_wasmer::Store,
+) -> core::result::Result<Module, CachedModuleErr> {
```
Comment (gshep) [inline sandbox/host/src/sandbox/wasmer_backend/cache.rs:54]: `Result` can be imported

File: sandbox/host/src/sandbox/wasmer_backend/cache.rs [hunk 5]
```diff
@@ -0,0 +1,211 @@
+use std::{
```
Comment (gshep) [inline sandbox/host/src/sandbox/wasmer_backend/cache.rs:1]: the license in the head is missing

**Review by gshep — APPROVED** [review]
Good job!

File: sandbox/host/src/sandbox/wasmer_backend/cache.rs (hunk 3 above)
Comment (breathx) [reply sandbox/host/src/sandbox/wasmer_backend/cache.rs:119]: It will has additional useless info then?

**Review by breathx — APPROVED** [review]

File: sandbox/host/src/sandbox/wasmer_backend/cache.rs (hunk 1 above)
Comment (ByteNacked) [reply sandbox/host/src/sandbox/wasmer_backend/cache.rs]: Extracted

File: sandbox/host/src/sandbox/wasmer_backend/cache.rs (hunk 3 above)
Comment (ByteNacked) [reply sandbox/host/src/sandbox/wasmer_backend/cache.rs]: Removed misleading result in the log message

File: sandbox/host/src/sandbox/wasmer_backend/cache.rs (hunk 4 above)
Comment (ByteNacked) [reply sandbox/host/src/sandbox/wasmer_backend/cache.rs]: Fixed

File: sandbox/host/src/sandbox/wasmer_backend/cache.rs (hunk 5 above)
Comment (ByteNacked) [reply sandbox/host/src/sandbox/wasmer_backend/cache.rs:21]: Fixed

File: sandbox/host/src/sandbox/wasmer_backend/cache.rs (hunk 2 above)
Comment (ByteNacked) [reply sandbox/host/src/sandbox/wasmer_backend/cache.rs]: Added

**Review by ark0f — CHANGES_REQUESTED** [review]

File: sandbox/host/src/sandbox/wasmer_backend/cache.rs [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+    let res = fs_cache.store(code_hash, &serialized_module);
+    log::trace!("Store module in FS cache with result: {:?}", res);
+}
+
+/// Altered copy of the `FileSystemCache` struct from `wasmer_cache` crate.
+#[derive(Debug, Clone)]
+pub struct FileSystemCache {
+    path: PathBuf,
+    ext: Option<String>,
```
[the remark this answers is not in the document]
Comment (ByteNacked) [reply sandbox/host/src/sandbox/wasmer_backend/cache.rs]: Fixed

File: sandbox/host/src/sandbox/wasmer_backend/cache.rs [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+}
+
+impl FileSystemCache {
+    /// Construct a new `FileSystemCache` around the specified directory.
+    fn new<P: Into<PathBuf>>(path: P) -> io::Result<Self> {
```
[the remark this answers is not in the document]
Comment (ByteNacked) [reply sandbox/host/src/sandbox/wasmer_backend/cache.rs]: Removed redundant checks

File: sandbox/host/src/sandbox/wasmer_backend/cache.rs [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+        lru_lock.insert(CachedModule {
+            wasm: wasm.to_vec(),
+            serialized_module: serialized_module.clone(),
+        });
+
+        // SAFETY: We trust the module in FS cache.
+        let module = unsafe {
+            Module::deserialize(store, serialized_module).map_err(|_| {
+                // Module in FS cache is corrupted, remove it.
```
[the remark this answers is not in the document]
Comment (ByteNacked) [reply sandbox/host/src/sandbox/wasmer_backend/cache.rs]: Added

**Review by ark0f — APPROVED** [review]

**Pull request merged**